log/handler.go

```diff
@@ -1,375 +1,290 @@
 }
 
-// FailoverHandler writes all log records to the first handler
-// specified, but will failover and write to the second handler if
-// the first handler has failed, and so on for all handlers specified.
-// For example you might want to log to a network socket, but failover
-// to writing to a file if the network fails, and then to
-// standard out if the file write fails:
+// TerminalHandler returns a handler which formats log records at all levels optimized for human readability on
+// a terminal with color-coded level output and terser human friendly timestamp.
+// This format should only be used for interactive programs or while developing.
 //
-//	log.FailoverHandler(
-//	    log.Must.NetHandler("tcp", ":9090", log.JSONFormat()),
-//	    log.Must.FileHandler("/var/log/app.log", log.LogfmtFormat()),
-//	    log.StdoutHandler)
+//	[LEVEL] [TIME] MESSAGE key=value key=value ...
 //
-// All writes that do not go to the first handler will add context with keys of
-// the form "failover_err_{idx}" which explain the error encountered while
-// trying to write to the handlers before them in the list.
-func FailoverHandler(hs ...Handler) Handler {
-	return FuncHandler(func(r *Record) error {
-		var err error
-		for i, h := range hs {
-			err = h.Log(r)
-			if err == nil {
-				return nil
-			}
-			r.Ctx = append(r.Ctx, fmt.Sprintf("failover_err_%d", i), err)
-		}
-
-		return err
-	})
+// Example:
+//
+//	[DBUG] [May 16 20:58:45] remove route ns=haproxy addr=127.0.0.1:50002
+func TerminalHandler(wr io.Writer, useColor bool) slog.Handler {
+	return TerminalHandlerWithLevel(wr, levelMaxVerbosity, useColor)
 }
 
-// ChannelHandler writes all records to the given channel.
-// It blocks if the channel is full. Useful for async processing
-// of log messages, it's used by BufferedHandler.
-func ChannelHandler(recs chan<- *Record) Handler {
-	return FuncHandler(func(r *Record) error {
-		recs <- r
-		return nil
-	})
+// TerminalHandlerWithLevel returns the same handler as TerminalHandler but only outputs
+// records which are less than or equal to the specified verbosity level.
+func TerminalHandlerWithLevel(wr io.Writer, lvl slog.Level, useColor bool) slog.Handler {
+	return &terminalHandler{
+		sync.Mutex{},
+		wr,
+		lvl,
+		useColor,
+		[]slog.Attr{},
+	}
 }
 
-// BufferedHandler writes all records to a buffered
-// channel of the given size which flushes into the wrapped
-// handler whenever it is available for writing. Since these
-// writes happen asynchronously, all writes to a BufferedHandler
-// never return an error and any errors from the wrapped handler are ignored.
-func BufferedHandler(bufSize int, h Handler) Handler {
-	recs := make(chan *Record, bufSize)
-	go func() {
-		for m := range recs {
-			_ = h.Log(m)
-		}
-	}()
-	return ChannelHandler(recs)
+func (h *terminalHandler) Handle(_ context.Context, r slog.Record) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	r.AddAttrs(h.attrs...)
+	h.wr.Write(TerminalFormat(r, h.useColor))
```

Comment:
Might be better to pass the `h.attrs` to `TerminalFormat`, instead of temporarily adding them to the record. The `TerminalFormat` internally calls `logfmt`, which _does_ have a parameter `commonAttrs []slog.Attr` where these extra attrs would fit perfectly (?)

Reply:
Yeah good idea.

Reply:
weird.  that extra parameter is unused anywhere.  the code that used it must have been deleted at some point...

Reply:
Yep, when you dropped `LogfmtFormat()`:

> If you also drop the method LogfmtFormat(), then TerminalFormat is the only caller of logfmt, so that could also be made into a function on terminalHandler, and suddenly both the fieldPadding and the handler's attrs are accessible for both these methods.